sparse/impl/KokkosSparse_sptrsv_solve_impl.hpp

```diff
@@ -100,550 +98,296 @@ struct SptrsvWrap {
-
-    const bool is_lowertri;
-
-    long node_count;  // like "block" offset into ngbl, my_league is the "local"
-                      // offset
-    long dense_nrows;
-
-    TriLvlSchedTP1SolverFunctorDiagValues(
-        const RowMapType &row_map_, const EntriesType &entries_,
-        const ValuesType &values_, LHSType &lhs_, const RHSType &rhs_,
-        const entries_t &nodes_grouped_by_level_,
-        const ValuesType &diagonal_values_, const bool is_lowertri_,
-        long node_count_, long dense_nrows_ = 0)
-        : row_map(row_map_),
-          entries(entries_),
-          values(values_),
-          lhs(lhs_),
-          rhs(rhs_),
-          nodes_grouped_by_level(nodes_grouped_by_level_),
-          diagonal_values(diagonal_values_),
-          is_lowertri(is_lowertri_),
-          node_count(node_count_),
-          dense_nrows(dense_nrows_) {}
 
-    KOKKOS_INLINE_FUNCTION
-    void operator()(const member_type &team) const {
-      auto my_league = team.league_rank();  // map to rowid
-      auto rowid     = nodes_grouped_by_level(my_league + node_count);
-      auto my_rank   = team.team_rank();
-
-      auto soffset   = row_map(rowid);
-      auto eoffset   = row_map(rowid + 1);
-      auto rhs_rowid = rhs(rowid);
-      scalar_t diff  = scalar_t(0.0);
-
-      Kokkos::parallel_reduce(
-          Kokkos::TeamThreadRange(team, soffset, eoffset),
-          [&](const long ptr, scalar_t &tdiff) {
-            auto colid = entries(ptr);
-            auto val   = values(ptr);
-            if (colid != rowid) {
-              tdiff = tdiff - val * lhs(colid);
-            }
-          },
-          diff);
+    template <bool IsSerial, bool IsSorted, bool IsLower,
+              bool UseThreadVec = false>
+    KOKKOS_INLINE_FUNCTION void solve_impl(const member_type *team,
+                                           const int my_rank,
+                                           const long node_count) const {
+      if (!IsSerial && BlockEnabled) {
+        KK_KERNEL_ASSERT_MSG(
+            !UseThreadVec,
+            "ThreadVectorRanges are not yet supported for block-enabled");
+      }
+      if (IsSerial) {
+        KK_KERNEL_ASSERT_MSG(!UseThreadVec,
+                             "Requested thread vector range in serial?");
+      }
+      const auto rowid   = nodes_grouped_by_level(my_rank + node_count);
+      const auto soffset = row_map(rowid);
+      const auto eoffset = row_map(rowid + 1);
+      const auto rhs_val = rhs(rowid);
+      scalar_t &lhs_val  = lhs(rowid);
+
+      // Set up range to auto-skip diag if is sorted
+      const auto itr_b = soffset + (IsSorted ? (IsLower ? 0 : 1) : 0);
+      const auto itr_e = eoffset - (IsSorted ? (IsLower ? 1 : 0) : 0);
+
+      // We don't need the reducer to find the diag item if sorted
+      using reducer_t =
+          std::conditional_t<IsSorted, ReduceSumFunctor, ReduceSumDiagFunctor>;
+      reducer_t rf{this, rowid, -1};
+
+      if (IsSerial) {
```

Comment:
Wow, all this time I thought compilers were smart enough to do this automatically. I may need to revisit some spiluk code as well.

Reply:
It will be smart enough to eliminate the unreachable branch during optimization, but before that it still has to be type checked. So it's a good practice to use constexpr when possible because it avoids redundant work for the compiler

Reply:
The other thing is it lets you put code in the branch that will only compile when the condition is true (like static_asserts, or functions that are only defined for certain types)

Reply:
Thanks for the clarification. I was panicking there for a second.